L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+        void ICollection<KeyValuePair<string, object>>.Add(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.Clear();
+        bool ICollection<KeyValuePair<string, object>>.Contains(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.CopyTo(KeyValuePair<string, object?>[] array, int arrayIndex);
+        bool ICollection<KeyValuePair<string, object>>.Remove(KeyValuePair<string, object?> item);
+        void IDictionary<string, object>.Add(string key, object? value);
+        bool IDictionary<string, object>.ContainsKey(string key);
+        object? IDictionary<string, object>.get_Item(string key);
+        bool IDictionary<string, object>.Remove(string key);
+        void IDictionary<string, object>.set_Item(string key, object? value);
+        bool IDictionary<string, object>.TryGetValue(string key, out object? value);
+        IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator();
+    }
+    public readonly struct BalancerAttributesKey<TValue> {
+        public BalancerAttributesKey(string key);
+        public string Key { get; }
+    }
+    public class BalancerState {
+        public BalancerState(ConnectivityState connectivityState, SubChannelPicker picker);
+        public ConnectivityState ConnectivityState { get; }
+        public SubChannelPicker Picker { get; }
+    }
+    public class ChannelState {
+        public ChannelState(AddressResolverResult resolverState, BalancerAttributes options);
+        public BalancerAttributes Options { get; set; }
+        public AddressResolverResult ResolverState { get; set; }
+    }
+    public class CompleteContext {
+        public CompleteContext();
+        public Exception? Error { get; set; }
+        public Metadata? Trailers { get; set; }
+    }
+    public class DnsAddressResolver : AddressResolver {
+        public DnsAddressResolver(Uri address, ILoggerFactory loggerFactory);
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override IDisposable Subscribe(IObserver<AddressResolverResult> observer);
+        protected override void Dispose(bool disposing);
+    }
+    public class DnsAddressResolverFactory : AddressResolverFactory {
+        public DnsAddressResolverFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override AddressResolver Create(Uri address, AddressResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
+        protected LoadBalancer();
+        public void Dispose();
+        public abstract void ResolverError(Exception exception);
+        public abstract void UpdateChannelState(ChannelState state);
+        public abstract void UpdateSubChannelState(SubChannel subChannel, SubChannelState state);
```

Comment:
As discussed above, this method should not be needed.

Reply:
Why won't it be needed? It is used to notify the LB that a subchannel state has changed.

https://github.com/grpc/grpc-go/blob/3dd75a6888ce5d1b5195c5cf72241d9e36f68e42/balancer/balancer.go#L321-L323

Reply:
There definitely needs to be *some* way to notify the LB policy of connectivity state changes for subchannels.  The question is whether it makes more sense to have a method on the LB policy for the channel to invoke or to have a way of doing this that works through the subchannel object itself.  C-core does the latter, but Go (or Java) may do something similar to what you have here.

If @dfawley is happy with how this API works in Go today, this may be fine.  @ejona86 may also be able to weigh in with how Java's API works.

Reply:
I've updated the design to remove this method and use a subchannel watcher to get status updates. This matches what Java, Node, and C++ do.